### PR TITLE
Add mypy info to Certbot docs

### LINKED ISSUE #5885: Write up info on mypy and typing in Certbot
Do this before Pycon (May 9)

### PR DESCRIPTION
**Pull request by ohemorange**
Fixes #5885.

Here's what it looks like:

screenshot from 2018-05-23 12-41-50

### CODE REVIEW

File: docs/contributing.rst [hunk 1]
```diff
@@ -312,6 +312,33 @@ Please:
 .. _PEP 8 - Style Guide for Python Code:
   https://www.python.org/dev/peps/pep-0008
 
+Mypy type annotations
+=====================
+
+Certbot uses the ``mypy`` static type checker. Python 3 natively supports official type annotations, which can then be tested for consistency using ``mypy``. Python 3 doesn’t, but type annotations can be `added in comments`_. ``mypy`` does some type checks even without type annotations; we can find bugs in Certbot even without a fully annotated codebase.
```
Comment (bmw) [inline docs/contributing.rst]: nit: I think it'd be cool if at least the first usage of `mypy` here linked to the mypy documentation, but this is an improvement we can always add later.

Comment (bmw) [reply docs/contributing.rst]: s/Python 3 doesn't/Python 2 doesn't

(a find replace command like that probably needs some backslashes for the apostrophes, but I think your ability to parse that is probably better than a command like `sed`.

File: docs/contributing.rst [hunk 2]
```diff
@@ -312,6 +312,33 @@ Please:
...
+
+Certbot supports both Python 2 and 3, so we’re using Python 2-style annotations.
+
+Zulip wrote a `great guide`_ to using mypy. It’s useful, but you don’t have to read the whole thing to start contributing to Certbot.
+
+To run ``mypy`` on Certbot, use ``tox -e mypy`` on a machine that has Python 3 installed.
+
+Note that instead of just importing ``typing``, due to packaging issues, in Certbot we use the following import statement:
```
Comment (bmw) [inline docs/contributing.rst]: nit: "...in Certbot we import from acme.magic_typing and have to add some comments for pylint like this:"

My thinking is just to expand and comment on what's actually noteworthy here.

File: docs/contributing.rst [hunk 3]
```diff
@@ -312,6 +312,33 @@ Please:
...
+  from acme.magic_typing import Dict # pylint: disable=unused-import, no-name-in-module
+
+Also note that OpenSSL, which we rely on, has type definitions for crypto but not SSL. We use both. Those imports should look like this:
+
+.. code-block:: python
+
+  from OpenSSL import crypto
+  from OpenSSL import SSL # type: ignore # https://github.com/python/typeshed/issues/2052
+
+.. _added in comments: http://mypy.readthedocs.io/en/latest/cheat_sheet.html
```
Comment (bmw) [inline docs/contributing.rst]: https

File: docs/contributing.rst [hunk 4]
```diff
@@ -312,6 +312,33 @@ Please:
...
+.. _great guide: http://blog.zulip.org/2016/10/13/static-types-in-python-oh-mypy/
```
Comment (bmw) [inline docs/contributing.rst]: https

**Comment by ohemorange** [comment]
screenshot from 2018-05-31 13-34-11

**Pull request merged**